x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/DomainSplitFunction.java

```diff
@@ -5,279 +5,285 @@
-                "        if (host == '.') {\n" +
-                "            return ['',''];\n" +
-                "        }\n" +
-                "        return ['', host];\n" +
-                "    }\n" +
-                "    def normalizedHost = host;\n" +
-                "    normalizedHost = normalizedHost.toLowerCase();\n" +
-                "    List parts = splitDomain(normalizedHost);\n" +
-                "    int publicSuffixIndex = findPublicSuffix(params, parts);\n" +
-                "    if (publicSuffixIndex == 0) {\n" +
-                "        return ['', host];\n" +
-                "    }\n" +
-                "    String highestRegistered = '';\n" +
-                "    /* for the case where the host is internal like .local so is not a recognised public suffix */\n" +
-                "    if (publicSuffixIndex == -1) {\n" +
-                "        if (!parts.isEmpty()) {\n" +
-                "            if (parts.size() == 1) {\n" +
-                "                return ['', host];\n" +
-                "            }\n" +
-                "            if (parts.size() > 2) {\n" +
-                "                boolean allNumeric = true;\n" +
-                "                String value = parts.get(parts.size() - 1);\n" +
-                "                for (int i = 0; i < value.length(); i++) {\n" +
-                "                    if (!Character.isDigit(value.charAt(i))) {\n" +
-                "                        allNumeric = false;\n" +
-                "                        break;\n" +
-                "                    }\n" +
-                "                }\n" +
-                "                if (allNumeric) {\n" +
-                "                    highestRegistered = parts.get(parts.size() - 2) + '.' + parts.get(parts.size() - 1);\n" +
-                "                } else {\n" +
-                "                    highestRegistered = parts.get(parts.size() - 1);\n" +
-                "                }\n" +
-                "\n" +
-                "            } else {\n" +
-                "                highestRegistered = parts.get(parts.size() - 1);\n" +
-                "            }\n" +
-                "        }\n" +
-                "    } else {\n" +
-                "        /* HRD is the top private domain */\n" +
-                "        highestRegistered = topPrivateDomain(normalizedHost, parts, publicSuffixIndex);\n" +
-                "    }\n" +
-                "    String subDomain = host.substring(0, host.length() - highestRegistered.length());\n" +
-                "    if (subDomain.endsWith('.')) {\n" +
-                "        subDomain = subDomain.substring(0, subDomain.length() - 1);\n" +
-                "    }\n" +
-                "    return [subDomain, highestRegistered];\n" +
-                "}\n";
-        fn = fn.replace("\n"," ");
-        function = fn;
+    private static List<String> split(String value) {
```

Comment:
Is this not the same as `value.split("\.")` baring the fact that returns an array and this a list?

Reply:
I did not modify the code at all. I only extracted it from the string prepended to painless.

Reply:
I'm happy to do a cleanup pass through this once there are unit tests, if you are able to find those.

Reply:
I thought it might be because painless does not support arrays but that isn't the [case](https://www.elastic.co/guide/en/elasticsearch/painless/master/painless-operators-array.html). @polyfractal any ideas why?

Reply:
Yeah, no idea.  Painless was pretty new when ML integrated, so it may be a new(er) thing... more likely it was just an oversight.  At the time I was trying hard not to modify the original as it was transliterated into painless, so could have been an artifact of that.

:man_shrugging: